Mathlib/Topology/EMetricSpace/PairReduction.lean

```diff
@@ -0,0 +1,427 @@
+/-
+Copyright (c) 2025 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne, David Ledvinka
+-/
+import Mathlib.Analysis.SpecialFunctions.Log.ENNRealLogExp
+import Mathlib.Order.CompletePartialOrder
+
+/-!
+# Pair Reduction
+
+The goal of this file is to prove the theorem `pair_reduction` which is used to prove a
+Kolmogorov–Chentsov theorem for general metric spaces. Given pseudometric spaces `T` and `E`,
+`c ≥ 0`, and a finite subset `J` of `T` such that `|J| ≤ aⁿ` for some `a ≥ 0` and `n : ℕ`,
+`pair_reduction` states that there exists a set `K ⊆ J²` such that for any function `f : T → E`:
+
+1. `|K| ≤ a|J|`
+2. `∀ s t ∈ K, d(s, t) ≤ cn`
+3. `sup_{s, t ∈ J : d(s, t) ≤ c} d(f(s), f(t)) ≤ 2 sup_{(s, t) ∈ K} d(f(s), f(t))`
+
+The key point being that it reduces bounding a supremum over points "close" in `J²` (which can be a
+set of size up to `|J|²`) to bounding a supremum over a set of points with size linear in `|J|`
+(whose points are still "close").
+
+This file was written with the goal of being used in the proof of a Kolmogorov–Chentsov theorem.
+In that application `f(s)` is random and our goal is to find an upper bound on the expectation
+of `sup_{s, t ∈ J : d(s, t) ≤ c} d(f(s), f(t))`.
+The way we deal with the expectation of a supremum is to bound it by a sum over the pairs.
+Suppose for simplicity that for all `u`, if `d(s, t) ≤ u` then `𝔼[d(f(s), f(t))] ≤ u`.
+If we work directly on the sum over all pairs we get a final upper bound `|J|² c`.
+However with the pair reduction lemma we get `2 𝔼[sup_{(s, t) ∈ K} d(f(s), f(t))] ≤ 2 a |J| c n`,
+and since `n = log|J|` we went from `|J|²` to `|J| log|J|`.
+
```

Comment:
Could you sketch the proof here, in an `## Implementation` section, and reference to a few of the lemmas below, to give your reader a hint of what is going on?